torch/csrc/dynamo/guards.cpp

```diff
@@ -664,6 +664,615 @@ static struct PyModuleDef _module = {
     -1,
     _methods};
 
+/**
+ * Stores relevant guard debug information, e.g., failure str for a LeafGuard
+ * failure. The data structure is also accessible in Python.
+ */
+
+class GuardDebugInfo {
+ public:
+  GuardDebugInfo(
+      bool result,
+      py::list verbose_code_parts,
+      int num_guards_executed)
+      : result(result),
+        verbose_code_parts(verbose_code_parts),
+        num_guards_executed(num_guards_executed) {}
+
+  // This constructor is used when guard succeeds.
+  GuardDebugInfo(bool result, int num_guards_executed)
+      : result(result), num_guards_executed(num_guards_executed) {}
+
+  GuardDebugInfo(
+      bool result,
+      std::string failed_reason,
+      int num_guards_executed)
+      : GuardDebugInfo(result, num_guards_executed) {
+    verbose_code_parts.append(failed_reason);
+  }
+
+  std::string to_string() {
+    std::stringstream ss;
+    ss << "GuardDebugInfo(\n"
+       << "result=" << result << ",\n"
+       << "verbose_code_parts=" << verbose_code_parts << ",\n"
+       << "num_guards_executed=" << num_guards_executed << ")\n";
+    return ss.str();
+  }
+
+  // Whether the guard passed or failed.
+  bool result;
+
+  // This is a list of verbose_code_parts for the failed guard. When there are
+  // more than one verbose_code_parts, then recompilation reasoning infra on the
+  // Python side can iterate over this list and eval each string to pinpoint the
+  // exact code part that failed.
+  py::list verbose_code_parts;
+
+  // Total number of executed guards so far. This is helpful in debugging if
+  // shuffling is working.
+  int num_guards_executed;
+};
+
+/**
+ * Base class for the leaf guard in the GuardManager hierarchy.
+ */
+class LeafGuard {
+ public:
+  LeafGuard(py::object verbose_code_parts)
+      : _verbose_code_parts(verbose_code_parts) {}
+
+  // check function could be called from python. This is useful for debugging
+  // purpose.
+  bool check(py::handle value) {
+    return check_nopybind(value.ptr());
+  }
+
+  GuardDebugInfo check_verbose(py::handle value) {
+    return check_verbose_nopybind(value.ptr());
+  }
+
+  virtual GuardDebugInfo check_verbose_nopybind(
+      PyObject* value) { // borrowed ref
+    bool result = check_nopybind(value);
+    if (!result) {
+      return GuardDebugInfo(result, _verbose_code_parts, 0);
+    }
+    return GuardDebugInfo(true, 0);
+  }
+
+  py::list verbose_code_parts() {
+    return _verbose_code_parts;
+  }
+
+  // This is on the hot path and avoids any refcounting code from pybind. This
+  // is not exposed to Python and can only be called from C++.
+  virtual bool check_nopybind(PyObject* value) = 0;
+  virtual ~LeafGuard() = default;
+
+ private:
+  // This is set while constructing the leaf guard. This is used for identifying
+  // the cause of recompilation.
+  py::list _verbose_code_parts;
```

Comment:
Its a list of strings. Usage is something like

```
guard = guards.TYPE_MATCH(id_type(foo), ["type(x) == int"])
```

Its a list because we can have multiple `verbose_code_parts` for to the same guard. So, I use `py::object` to hold a reference for the list of strings.